Separate Python interpreter from inspector and pass selected python environment down through deployment

sagerb opened the pull request:
<!-- Provide a general summary of your changes in the title. -->
<!-- Examples: Updates pull request template -->

## Intent

<!-- Describe what problem you are addressing in this pull request. -->
<!-- If this change is associated with an open issue, please link to it here. -->
<!-- Example: "Resolves #24" -->
<!-- See https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue -->

This PR includes the functionality required to query and pass on the selected Python interpreter path active within VSCode or Positron and passes that down through the applicable APIs for reference within the agent. The agent then uses the selected Python interpreter when needed.

Resolves #2494 
Resolves #2496

## Type of Change

<!-- What types of changes does your code introduce? Put an `x` in all the boxes that apply. -->
<!-- If you check more than one box, you may need to refactor this change into separate pull requests -->

- - [ ] Bug Fix <!-- A change which fixes an existing issue -->
- - [x] New Feature <!-- A change which adds additional functionality -->
- - [ ] Breaking Change <!-- A breaking change which causes existing functionality to change -->
- - [ ] Documentation <!-- User or developer documentation -->
- - [x] Refactor <!-- Code restructuring -->
- - [ ] Tooling <!-- Build, CI, or release scripts and configuration -->

## Approach

<!-- Describe how you solved this problem and any trade-offs you encountered. -->
<!-- Link any additional documentation associated with this change here -->

To accomplish having the availability of the Python active interpreter within the agent, I followed the pattern established when we refactored the R interpreter-related functionality.

This involved a number of incremental steps, which have been grouped into related commits:
- [Moved ExistsFunc type definition into common utils](https://github.com/posit-dev/publisher/commit/32c68818a23509af02423d0b76a905f105fbe8a7) & [Moved ExistsFunc type definition into common utils #2](https://github.com/posit-dev/publisher/commit/35b50968c8969eeb2501c816ac6385f74e27ca42)
- [Refactored Positron API access to handle both R and Python selections](https://github.com/posit-dev/publisher/commit/f93c1eaaeb59708d88ae53da0a203b1b7b776f09)
- [Update Deployment API to send in selected python interpreters. Also updates commands which are unused, but have not yet been removed.](https://github.com/posit-dev/publisher/commit/0aa11593a818b9b95b127ed778ab589e978ff34f) 
- [Create the new python interpreter (and the associated test & mock)](https://github.com/posit-dev/publisher/commit/35112d3911bd953145d7c55aa73527c9c6d9b687)
- [Update the python inspectors to now use the new python interpreter](https://github.com/posit-dev/publisher/commit/8c43d77faa8ee8a783f921a206b21a1dc302bbd3)
- [Move Requirements file functionality into the python dependencies package (pydeps). Also updates code to use the new Python Inspector call signature.](https://github.com/posit-dev/publisher/commit/02e3c063450fb8eba89fd6a00d875df045ab1a43) 
- [Handle the changes required for the new python interpreter and it's impact on the python inspector.](https://github.com/posit-dev/publisher/commit/18014ae61fdcf9f33dc8aab47876b7ec5960cda5) 
- [Add test for multiple types of configs from initialize/inspect](https://github.com/posit-dev/publisher/commit/5ce683a1cd184efff12dc60f0034ccca7ed0e0dc)

## User Impact

<!-- Describe how this change will affect users. -->
<!-- Specifically changes to the flow and trade-offs we are making. -->

The publisher now recognizes the active python interpreter as selected within Microsoft's Python extension and Posit's Positron IDE. This leads to a more consistent and controllable environment for the users.

## Automated Tests

<!-- Describe the automated tests associated with this change. -->
<!-- If automated tests are not included in this change, please state why. -->

Unit tests have been updated to reflect the changes.

## Directions for Reviewers

<!-- Provide steps for reviewers to validate this change manually. -->

I have grouped types of changes into individual commits. Reviewing them in that manner should help facilitate a more productive review of the many changes. 

The bulk of the functionality change involves creating the new Python interpreter object, extracting this Python interpreter functionality out of the inspection package, and then assessing the impact that those changes had on the other code (especially on the initialize package functionality). In addition, some changes were then needed to extract the correct values from the VSCode/Positron environment and pass them through the API to be available within the API handlers.

The changes involved in this PR are best verified by inspection of our different content types / sample projects. There should be no difference in the detection of Python or not, but the version string will now be consistent with the active version of Python that the user is running (assuming that they are setting this through one of the facilities we support - VSCode Python extension, Positron Python selection or first location found searching the path).

## Checklist

<!--- Go over all the following points, and put an `x` in all the boxes that apply: -->
<!--- If you need clarification on any of these, feel free to ask. We're here to help! -->

- [ ] I have updated [CHANGELOG.md](../CHANGELOG.md) to cover notable changes.

dotNomad commented on cmd/publisher/commands/requirements_create.go:
```diff
@@ -34,12 +35,12 @@ func (cmd *CreateRequirementsCommand) Run(args *cli_types.CommonArgs, ctx *cli_t
 		return err
 	}
-	err = inspector.WriteRequirementsFile(reqPath, reqs)
+	err = pydeps.WriteRequirementsFile(reqPath, reqs, util.NewAbsolutePath("bogus", nil))
```
Oh this is in the commands directory so this is part of the CLI. Feel free to ignore.

sagerb commented:
>your input may be valuable here to understand just how difficult either is to get this passing.

@dotNomad Thanks for diving in. I have just pushed a commit to remove that content type from the test and I have also updated that bogus string I added to be a bit more understandable.

We'll see if these changes now help the BAT tests to pass!

dotNomad commented:
> Thanks for diving in. I have just pushed a commit to remove that content type from the test and I have also updated that bogus string I added to be a bit more understandable.

Still failing. I think this is because the requirements file is also being generated for the `fastapi-simple` project and is empty given the dummy string.

dotNomad commented on internal/inspect/dependencies/pydeps/project_dependencies.go:
```diff
@@ -50,3 +57,38 @@ func (s *defaultDependencyScanner) ScanDependencies(base util.AbsolutePath, pyth
 }
+
+func GetRequirementsFilePath(base util.AbsolutePath) (util.RelativePath, bool, error) {
+	filePath := "requirements.txt"
```
I would include a comment here to note that this is used in inspection to find a `requirements.txt`

When reading this I was wondering why it wasn't using the config's package file filename briefly before looking at its usage.

sagerb replied:
I've renamed the function to `DoesDefaultRequirementsExist` as well as adding a comment.

dotNomad reviewed (approved):
I did some changes to my machine to test some different cases:

I removed my global Python and this PR still worked perfectly based on the Python interpreter I had selected in the Python VS Code extension.

I also got the correct error if I didn't have the Python extension or Python globally.

I tested changing the selected Python interpreter in Positron and got the expected configurations.

---

I went through this with a pretty fine tooth comb, both for testing and the review itself. I test on VS Code, Positron, with and without Python environments selected, with an entrypoint at the workspace level and below.

This is a fantastic refactor. I only had one comment about adding a comment where I was a bit confused on the intention. 🚀

dotNomad commented on .github/workflows/pull-request.yaml:
```diff
@@ -34,10 +34,10 @@ jobs:
-    needs: build
-    secrets: inherit
-    uses: ./.github/workflows/bats.yaml
+  # bats:
+  #   needs: build
+  #   secrets: inherit
+  #   uses: ./.github/workflows/bats.yaml
```
We may want to do this in another PR for easier tracking. Maybe hold on merging this until we make an effort to remove.

dotNomad replied:
https://github.com/posit-dev/publisher/pull/2558 was opened to remove these so we can solve the conflict with whatever merges first.
